Reader fix

pjvanwijk opened the pull request:
Er kunnen nu meerdere files worden ingelezen met de xml. ook excelfiles met eerste row als header

pjvanwijk committed:
handle eof by last empty row in excel file

pjvanwijk committed:
forgot to add columns to builder with firstrowheader property

pjvanwijk committed:
change deprecated date methods to gregorian calendar values

pjvanwijk committed:
modify null values in test file

louisgosschalk commented:
mooi klein, ziet er prima uit

JensLangerak commented on src/main/java/model/input/file/ExcelFile.java:
```diff
@@ -40,42 +39,69 @@ protected DataTable createTable(Iterator<Row> rowIterator) throws FileNotFoundEx
 			}
 		}
-		
+
```
kun je deze methode hier splitsen?

JensLangerak commented on src/main/java/model/input/file/ExcelFile.java:
```diff
@@ -40,42 +39,69 @@ protected DataTable createTable(Iterator<Row> rowIterator) throws FileNotFoundEx
+					calendar.get(Calendar.SECOND));
+		} else {
+			double cellValue = cell.getNumericCellValue();
+			value = (cellValue % 1 == 0)
```
Je krijgt nu columns waar zowel floats als ints in kunnen staan

Boudewijn26 commented:
Je gebruikt nog steeds de map waar ik in #135 al op was afgeknapt. Aangezien de branch reader-fix heet, zou ik het wel beter vinden om dat ook te fixen.

pjvanwijk commented:
Ik close hem nu tot ik de comments heb verwerkt

pjvanwijk committed:
Merge branch 'master' of https://github.com/clanghout/Health-Informatics-3 into reader-fix

Merging for feature count-from-end

pjvanwijk committed:
Merge branch 'master' of https://github.com/clanghout/Health-Informatics-3 into reader-fix

Conflicts:
	src/main/java/model/input/file/ExcelFile.java

pjvanwijk committed:
Refactor input to use columninfo object instead of a map

pjvanwijk committed:
add parsing functionality for excelfile

pjvanwijk committed:
parsing of datevalues timevalues

pjvanwijk committed:
add parsing for  more date and time types

pjvanwijk committed:
add excel file tests

pjvanwijk committed:
add tests and fix the faults

now datevalues and timevalues and datetimevalues can be parsed from strings in plaintextfiles.
Date formatted cells in excel files can also be converted to our values

pjvanwijk commented:
De reader werkt nu met een columninfo class die de parse info bevat hoe de data in de kolom geparst moet worden. Nu is het dan ook mogelijk om dates en times en datetimes te parsen door de parse format mee te geven als attribute in een column element van de xml

pjvanwijk committed:
add more tests and included date format in the xml parser

pjvanwijk committed:
remove unused imports and add javadoc

pjvanwijk committed:
fix checkstyle pmd findbugs

JensLangerak commented on src/main/java/model/input/file/DataFile.java:
```diff
@@ -166,7 +163,7 @@ public static Class getColumnType(String type) {
 	 */
-	public static String getStringColumnType(Class type) {
+	public static String getStringColumnType(Class<? extends DataValue> type) {
 		if (type == IntValue.class) {
```
boolean value ontbreekt

JensLangerak commented on src/main/java/model/input/file/DataFile.java:
```diff
@@ -311,32 +257,84 @@ public void createMetaDataValue(String name, String type) {
+	 * @return The DataValue representing the null
+	 */
+	public static DataValue createNullValue(Class<? extends DataValue> type) {
+		if (type == StringValue.class) {
```
DataValue.getNullInstance(type);

JensLangerak commented on src/main/java/model/input/file/DataFile.java:
```diff
@@ -311,32 +257,84 @@ public void createMetaDataValue(String name, String type) {
-
-			this.metaDataValue = parseDataValue(metaValue, type);
+			Class<? extends DataValue> typeClass = DataFile.getColumnType(type);
+			this.metaDataValue = parseSimpleDataValue(metaValue, typeClass);
```
waarom simpleDataValue. Als ik die methode bekijk kan die alleen getallen of strings zijn. Waarom(of waar als het wel kan) sta je niet alle types toe?

pjvanwijk replied:
Simpledatavalue is inderdaad alleen maar string of getallen. Deze kunnen in alle gevallen (txt of excel) naar een waarde geparst worden. Maar dates zijn afhankelijk van de type file waarin ze voorkomen. In een textfile moeten ze worden geparst met een format en in een excelfile kan je een datecell door POI een date laten returnen

JensLangerak replied:
ja dat had ik gezien maar ik was niet helemaal duigelijk. Ik doelde hier op de metadata. Want volgens mij kan de metatdata nu geen datum zijn.

JensLangerak commented on src/main/java/model/input/file/DataFile.java:
```diff
@@ -311,32 +257,84 @@ public void createMetaDataValue(String name, String type) {
-				return new StringValue(value);
-			default:
-				throw new RuntimeException("Class has not yet implemented");
+	public static DataValue parseSimpleDataValue(String value, Class<? extends DataValue> type) {
```
zie ook mijn comment bij plaintextfile, maar je wilt niet drie verschillende publieke methodes hebben voor het bepalen de type. Je wilt een publieke methode die voor alle type kan worden aangeroepen. Die methode mag wel verschillende andere private methodes aanroepen

JensLangerak commented on src/main/java/model/input/file/PlainTextFile.java:
```diff
@@ -127,30 +127,20 @@ private void addRowsToBuilder(List<String> lines) {
+	private DataValue toDataValue(String value, ColumnInfo columnInfo) {
+		if (value.isEmpty()) {
+			return createNullValue(columnInfo.getType());
+		} else if (columnInfo.getType() == DateValue.class) {
```
het is raar om hier slechts te controleren op de helft van de types. Je moet of alle type controleren en dan hier per type de juiste value aan maken, of je moet een methode aanroepen die zelf kan bekijken wat hij voor dat type moet doen.

JensLangerak commented on src/main/java/model/input/file/ExcelFile.java:
```diff
@@ -72,19 +84,53 @@ private void addRows(Iterator<Row> rowIterator) {
+		}
+	}
+
+	private DataValue parseNumValue(Cell cell, Class<? extends DataValue> type) {
```
ik kan het zelf nu niet zo makkelijk checken dus ik vraag het maar. Hou je rekening dat de numvalues voor excel datums anders is dan de numvalue van de calendar? de calendar gaat uit van de unix time stamp excel niet.

JensLangerak commented on src/main/java/model/input/file/ExcelFile.java:
```diff
@@ -72,19 +84,53 @@ private void addRows(Iterator<Row> rowIterator) {
+		if (type == DateTimeValue.class) {
+			Instant instant = Instant.ofEpochMilli(date.getTime());
+			return new DateTimeValue(LocalDateTime.ofInstant(instant, ZoneId.systemDefault()));
+		} else
```
else if graag op een regel houden

JensLangerak commented on src/main/java/model/input/file/PlainTextFile.java:
```diff
@@ -43,15 +45,14 @@ public DataTable createDataTable() throws IOException {
 				String[] sections = headers.split(delimiter);
-				for (int i = 0; i < getColumnTypes().length; i++) {
-					getColumns().put(sections[i], getColumnTypes()[i]);
+				for (int i = 0; i < getColumns().size(); i++) {
```
het is een vrij lange methode zonder line breaks wat hem erg slecht leesbaar maakt. Dus je wilt deze methode inkorten. Ik zou bijvoorbeeld de body van de if in een aparte methode gooien

pjvanwijk committed:
change nullcreation to use existing method

pjvanwijk committed:
fix github comments for readability

pjvanwijk committed:
add boolvalue to recognized columntypes

pjvanwijk committed:
Merge branch 'master' of https://github.com/clanghout/Health-Informatics-3 into reader-fix

pjvanwijk committed:
able to parse dates from stringcells in excelfiles

pjvanwijk committed:
able to parse exceldates with xml keyword exceldate

pjvanwijk committed:
Add NULL string cells to parser

pjvanwijk committed:
add boolvalue to parser

Boudewijn26 commented on src/main/java/model/input/file/DataFile.java:
```diff
@@ -311,35 +261,104 @@ public void createMetaDataValue(String name, String type) {
-				return new StringValue(value);
-			default:
-				throw new RuntimeException("Class has not yet implemented");
+	protected DataValue parseSimpleDataValue(String value, Class<? extends DataValue> type) {
```
Het zou waarschijnlijk mooier zijn als je deze methodes verplaatst naar ColumnInfo, maar dat is werk en nu hebben we niet echt tijd.

Boudewijn26 commented:
Het is niet de mooiste code, maar volgens mij werkt het.

pjvanwijk committed:
check on columninfo to parse instead of celltype

pjvanwijk committed:
remove calendar and use localtime in excelfile parsedatecellvalue

pjvanwijk committed:
add ability to give own value to metadata column

pjvanwijk committed:
fix checkstyle
